lib/hanami/application.rb

```diff
@@ -137,6 +140,23 @@ def booted?
         !!@booted # rubocop:disable Style/DoubleNegation
       end
 
+      def settings(&block) # rubocop:disable Metrics/MethodLength
+        if block
+          @_settings = Application::Settings.build(
+            configuration.settings_loader,
+            configuration.settings_loader_options,
+            &block
+          )
+        elsif instance_variable_defined?(:@_settings)
+          @_settings
+        else
+          # Load settings lazily so they can be used to configure the
+          # Hanami::Application subclass (before the application has inited)
+          load_settings
+          @_settings ||= nil
+        end
+      end
```

Comment:
FWIW I couldn't follow the `@_mutex.synchronize do ... end` dance that is followed in e.g. `.routes` inside this same class, because in the `else` case above, running `load_settings` will most likely result in the `.settings` method being called _again_ when `config/settings.rb` is loaded. This results in a deadlock if the code in this method is wrapped in the synchronize block.

The particular behaviour in that else case is important for the lazy loading of settings upon first access.

And look, while I'm here, I'll be honest: I don't _really_ understand why we're doing the synchronizing of things like the `@_routes` assignment inside this class... what is it actually in aid of? And if it's really necessary, how can we establish similar protections here?

Reply:
@jodosha Are you fine with this approach?

I wouldn't mind @flash-gordon's thoughts on this too, since he's said intelligent things about mutexes whenever I've asked him about them :)